rfc/rfc-0002-bindings.md

```diff
@@ -0,0 +1,97 @@
+# RFC-0002: Bindings
+
+**To be Reviewed By:** 
+
+**Authors:** @scothis
+
+**Status:** **Active** | Accepted | Dropped | Superseded
+
+**Pull Request URL:** #1360
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** N/A
+
+
+## Problem
+Non-trivial workloads often need to access resources, like databases, message brokers or any credentialed resource. In Cloud Foundry, these are called services and they are provisioned from a broker and then bound to applications at both build and runtime. As Kubernetes has a different notion of a Service, this RFC will refer to "service bindings" as simply bindings.
+
+The binding contains both metadata and credentials to provide a workload the details it needs to connect and consume the resource.
+
+### Anti-Goals
+A generic mechanism to provision bindings, and the specific structure of the metadata and credentials are out of scope for this RFC.
+
+## Solution
+The Cloud Native Buildpacks project has a [spec for Bindings](https://github.com/buildpack/spec/blob/master/extensions/bindings.md) which defines the form bindings should take at both build and runtime. In short, each binding is represented as a directory inside a well known directory with two sub-directories: `metadata` for general non-sensitive information about the binding, and `secret` for credentials and other sensitive information. Each attribute for the binding is a file in one of those two directories. The metadata is available at buildtime and both the metadata and secret is available at runtime.
+
+riff should adopt CNB Bindings spec for bindings at both build and runtime.
+
+Kubernetes uses [Volumes](https://kubernetes.io/docs/concepts/storage/volumes/) to inject files from an external source into a running container. While there is a diverse spectrum of volume implementations available, in order to maximize both compatibility with bog standard Kubernetes and simplify the implementation, we should focus on a [configMap volume](https://kubernetes.io/docs/concepts/storage/volumes/#configmap) for binding metadata and a [secret volume](https://kubernetes.io/docs/concepts/storage/volumes/#secret) for binding secrets, which map directly to Kubernetes [ConfigMaps](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/) and [Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) respectively.
+
+As a binding is backed by a ConfigMap and Secret a mechanism is needed to specify which ConfigMap and Secret. [ObjectReferences](https://godoc.org/k8s.io/api/core/v1#ObjectReference) are commonly used to define the coordinates to an arbitrary resources. A binding reference consists of metadata ObjectReference and a secret ObjectReference.
+
+Any resource that produces a binding should include the binding reference on it's status, for example:
```

Comment:
It's intentionally vague as we won't know the universe of resources that may be bindable.

One concrete example would be the Stream resource, as it has coordinates like the gateway address and topic (and someday credentials as well). As the Stream controller reconciles each Stream, it can create the binding and hold a reference to that binding on its status for consumers to discover. See https://github.com/projectriff/system/pull/181 as an early example.

I'm intentionally not proposing a Binding CRD, although I'm also not precluding such a resource from existing in the future that implements this duck type interface.

Reply:
It says "should", I assume that if there are resources that are used for binding and are not under our control, then they won't necessarily have this status.

Reply:
> I assume that if there are resources that are used for binding and are not under our control, then they won't necessarily have this status

I'm not trying to solve for the world at this point. We need to start somewhere and once we have a working foundation build out way out.